src/hp/derp/http/server.rs

```diff
@@ -1,50 +1,404 @@
+    pub fn request_handler(
+        mut self,
+        method: Method,
+        uri_path: &'static str,
+        handler: HyperFn,
+    ) -> Self {
+        self.handlers.insert((method, uri_path), handler);
+        self
+    }
+
+    /// Pass in a custom "404" handler.
+    pub fn not_found_handler(mut self, handler: HyperFn) -> Self {
+        self.not_found_fn = Some(handler);
+        self
+    }
+
+    /// Handle the derp endpoint in a custom way. This is required if no `SecretKey` was provided
+    /// to the builder.
+    pub fn derp_override(mut self, handler: HyperFn) -> Self {
+        self.derp_override = Some(handler);
+        self
+    }
+
+    /// Change the derp endpoint from "/derp" to `endpoint`.
+    pub fn derp_endpoint(mut self, endpoint: &'static str) -> Self {
+        self.derp_endpoint = endpoint;
+        self
+    }
+
+    /// Add http headers.
+    pub fn headers(mut self, headers: Headers) -> Self {
+        self.headers = headers;
+        self
+    }
+
+    /// Build and spawn an HTTP(S) derp Server
+    pub async fn spawn(self) -> Result<Server> {
+        ensure!(self.secret_key.is_some() || self.derp_override.is_some(), "Must provide a `SecretKey` for the derp server OR pass in an override function for the 'derp' endpoint");
+        let (derp_handler, derp_server) = if let Some(secret_key) = self.secret_key {
+            let server = crate::hp::derp::server::Server::new(secret_key, self.mesh_key);
+            println!("headers: {:?}", self.headers);
+            let header_map: HeaderMap = HeaderMap::from_iter(
+                self.headers
+                    .iter()
+                    .map(|(k, v)| (k.parse().unwrap(), v.parse().unwrap())),
+            );
+
+            (
+                DerpHandler::ConnHandler(server.client_conn_handler(header_map)),
+                Some(server),
+            )
+        } else {
+            (DerpHandler::Override(self.derp_override.unwrap()), None)
+        };
+        let h = self.headers.clone();
+        let not_found_fn = match self.not_found_fn {
+            Some(f) => f,
+            None => Box::new(move |mut res: ResponseBuilder| {
+                for (k, v) in h.iter() {
+                    res = res.header(*k, *v);
+                }
+                let r = res
+                    .status(StatusCode::NOT_FOUND)
+                    .body(b"Not Found"[..].into())
+                    .unwrap();
+                HyperResult::Ok(r)
+            }),
+        };
+
+        let service = DerpService::new(
+            self.handlers,
+            derp_handler,
+            self.derp_endpoint,
+            not_found_fn,
+            self.headers,
+        );
+
+        let server_state = ServerState {
+            addr: self.addr,
+            tls_config: self.tls_config,
+            server: derp_server,
+            service,
+        };
+
+        server_state.serve().await
+    }
+}
+
+impl std::fmt::Debug for ServerBuilder {
```

Comment:
I think dig went on a quest to remove the manual debug impls and introduced `derive_more::Debug`.  e.g. https://github.com/n0-computer/iroh/blob/6df66f479aac274f8231aaabbe3d11f07521b6a4/src/hp/magicsock/endpoint.rs#L80